pkg/reconciler/ingress/status_test.go

```
@@ -0,0 +1,301 @@
+		if r.Host != <-hosts {
+			w.WriteHeader(404)
+		}
+		requests <- struct{}{}
+	}))
+	defer ts.Close()
+	url, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("failed to parse URL %q: %v", ts.URL, err)
+	}
+
+	ready := make(chan *v1alpha3.VirtualService)
+	prober := NewStatusProber(
+		zaptest.NewLogger(t).Sugar(),
+		&fakeGatewayLister{
+			gateways: []*v1alpha3.Gateway{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "gateway",
+					},
+					Spec: v1alpha3.GatewaySpec{
+						Selector: map[string]string{
+							"gwt": "istio",
+						},
+					},
+				},
+			}},
+		&fakePodLister{
+			pods: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "gateway",
+					},
+					Status: v1.PodStatus{
+						PodIP: url.Host,
+					},
+				},
+			},
+		},
+		network.NewAutoTransport,
+		func(vs *v1alpha3.VirtualService) {
+			ready <- vs
+		})
+
+	prober.stateExpiration = 2 * time.Second
+	prober.cleanupPeriod = 500 * time.Millisecond
+
+	done := make(chan struct{})
+	defer close(done)
+	prober.Start(done)
+
+	// The first call to IsReady must succeed and return false
+	ok, err := prober.IsReady(vs)
+	if err != nil {
+		t.Fatalf("IsReady failed: %v", err)
+	}
+	if ok {
+		t.Fatalf("IsReady returned %v, want: %v", ok, false)
+	}
+
+	// Wait for the first request (failing) to be executed
+	<-requests
+
+	// Wait for the second request (success) to be executed
+	<-requests
+
+	// Wait for the probing to eventually succeed
+	<-ready
+
+	// The subsequent calls to IsReady must succeed and return true
+	for i := 0; i < 5; i++ {
+		ok, err = prober.IsReady(vs)
+		if err != nil {
+			t.Fatalf("IsReady failed: %v", err)
+		}
+		if !ok {
+			t.Fatalf("IsReady returned %v, want: %v", ok, false)
+		}
+
+		<-time.After(prober.cleanupPeriod)
+	}
+
+	// Wait for the cleanup to happen
+	<-time.After(prober.stateExpiration + prober.cleanupPeriod)
```

Move this into the select below so we fail faster?